server/src/utils/quest.js

```diff
@@ -6,44 +6,60 @@ import { Quest } from '@linux-odyssey/models'
 
 const questDirectory = path.join(process.cwd(), '..', 'quests')
 
-export default async function loadAndUpdateQuests() {
-  const questNames = await fs.readdir(questDirectory)
+async function mapQuests(callback) {
+  const questNames = await fs.readdir(questDirectory, {
+    withFileTypes: true,
+  })
 
-  const quests = Promise.all(
+  return Promise.all(
     questNames
-      .filter((name) => !name.startsWith('.'))
-      .map(async (id) => {
-        // Check if id is directory
-        const stat = await fs.stat(path.join(questDirectory, id))
-        if (!stat.isDirectory()) {
-          return null
-        }
-        const fullPath = path.join(questDirectory, id, 'game.yml')
-
-        try {
-          // Read and parse the README.md file.
-          const body = await fs.readFile(fullPath, 'utf-8')
-
-          const quest = yaml.parse(body, {
-            merge: true,
-          })
-
-          return Quest.findByIdAndUpdate(
-            id,
-            {
-              _id: id,
-              ...quest,
-            },
-            {
-              upsert: true,
-            }
-          )
-        } catch (error) {
-          console.error(`Error parsing quest ${id}:`, error)
-          throw error
-        }
+      .filter((dirent) => dirent.isDirectory() && !dirent.name.startsWith('.'))
+      .map((dirent) => {
+        const id = dirent.name
+        const questPath = path.join(questDirectory, id)
+        return callback(id, questPath)
       })
   )
+}
+
+export function loadAndUpdateQuests() {
+  return mapQuests(async (id, questPath) => {
+    const files = await fs.readdir(questPath)
+    if (!files.includes('game.yml')) {
+      throw new Error(`Quest ${id} is missing game.yml`)
+    }
+
+    try {
+      const body = await fs.readFile(path.join(questPath, 'game.yml'), 'utf-8')
+
+      const quest = yaml.parse(body, {
+        merge: true,
+      })
+
+      const image = files.includes('Dockerfile') ? id : 'base'
+
+      return Quest.findByIdAndUpdate(
+        id,
+        {
+          _id: id,
+          image,
+          ...quest,
+        },
+        {
+          upsert: true,
+        }
+      )
+    } catch (error) {
+      console.error(`Error parsing quest ${id}:`, error)
+      throw error
+    }
+  })
+}
 
-  return quests
+export async function getQuestDockerfiles() {
+  const quests = await mapQuests(async (id, questPath) => {
+    const files = await fs.readdir(questPath)
+    return files.includes('Dockerfile') ? { id, questPath } : null
+  })
+  return quests.filter((quest) => quest !== null)
```

Comment:
Do we get two return when calling this function?

Reply:
No, we only get one. The inner function inside map return a promise, and we use Promise.all to resolve all the promises at the same time, and finally return the result.